src/parser.rs

```diff
@@ -0,0 +1,202 @@
+            return Ok(self.advance().unwrap());
+        }
+
+        let peek = self.peek().unwrap();
+
+        return Err(ParseError { peek, message });
+    }
+
+    fn primary(&self) -> Expr {
+        //if (self.match_types(vec![TokenType::FALSE])) {
+        //    return Expr { Some(false)};
+        //}
+        //if (self.match_types(vec![TokenType::TRUE])) {
+        //    return Expr::Literal(Some(true));
+        //}
+        //if (self.match_types(vec![TokenType::NIL])) {
+        //    return Expr::Literal(None);
+        //}
+
+        //if (self.match_types(vec![TokenType::NUMBER, TokenType::STRING])) {
+        //    return Expr::Literal(self.previous().unwrap().literal);
+        //}
+
+        //if (self.match_types(vec![TokenType::LEFT_PAREN])) {
+        //    let expr = self.expression();
+        //    self.consume(
+        //        TokenType::RIGHT_PAREN,
+        //        String::from("Expect ')' after expression."),
+        //    );
+        //    return Expr::Grouping(expr);
+        //}
+        todo!();
+    }
+
+    fn unary<'a>(self) -> Expr<'a> {
+        if self.match_types(vec![TokenType::BANG, TokenType::MINUS]) {
+            let operator = self.previous();
+            let right = self.unary();
+            return Expr {
+                left: None,
+                operator,
+                right: Some(Box::new(right)),
+            };
+        }
+
+        return self.primary();
+    }
+
+    fn factor<'a>(mut self) -> Expr<'a> {
+        let mut expr = self.unary();
+
+        while self.match_types(vec![TokenType::SLASH, TokenType::STAR]) {
+            let operator = self.previous();
+            let right = self.unary();
+            expr = Expr {
+                left: Some(Box::new(expr)),
+                operator,
+                right: Some(Box::new(right)),
+            };
+        }
+
+        return expr;
+    }
+
+    fn term<'a>(self) -> Expr<'a> {
+        let mut expr = self.factor();
+
+        while self.match_types(vec![TokenType::MINUS, TokenType::PLUS]) {
+            let operator = self.previous();
+            let right = self.factor();
+            expr = Expr {
+                left: Some(Box::new(expr)),
+                operator,
+                right: Some(Box::new(right)),
+            }
+        }
+
+        return expr;
+    }
+
+    fn comparison<'a>(self) -> Expr<'a> {
+        let mut expr = self.term();
+
+        while self.match_types(vec![
+            TokenType::GREATER,
+            TokenType::GREATER_EQUAL,
+            TokenType::LESS,
+            TokenType::LESS_EQUAL,
+        ]) {
+            let operator = self.previous();
+            let right = self.term();
+            expr = Expr {
+                left: Some(Box::new(expr)),
+                operator,
+                right: Some(Box::new(right)),
+            }
+        }
+
+        return expr;
+    }
+
+    fn match_types(&mut self, types: Vec<TokenType>) -> bool {
```

Comment:
I guess having a vector here as well isn't great? I'm just doing it so I can pass in `n` many items. Could I do some sort of spread operator here instead?

Reply:
I think this perhaps over-abstracts when you could use `if !matches(foo, TokenType::Bar | ...)`.

If you'd really like to avoid the typing, you can either write methods to return the specific match types, e.g...

```rust
fn matches_literal(&self) -> bool {
  matches!(self,  Self::IDENTIFIER | Self::STRING | Self::NUMBER)
}

// Elsewhere:
if token.matches_literal() { ... }
```

Or write a macro... but again that might be over-engineering. At any rate here's how the above function could be a macro instead:

```rust
macro_rules! matches_literal {
  ($val: ident) => {
    matches!($val,  TokenType::IDENTIFIER | TokenType::STRING | TokenType::NUMBER)
  }
}

// Elsewhere:
if matches_literal!(token) { ... }
```

Reply:
> Could I do some sort of spread operator here instead?

Rust explicitly does not have variadic functions. If you want something to be variadic, you can only use macros.

Reply:
> I think this perhaps over-abstracts when you could use `if !matches(foo, TokenType::Bar | ...)`.

How should the `matches` function signature look if I did this?